flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherPauseResumeSplitReaderTest.java

```diff
@@ -0,0 +1,311 @@
+                fetcher.runOnce();
+            }
+        }
+    }
+
+    private static class MockSplitReaderUnsupportedPause extends MockSplitReader {
+        public MockSplitReaderUnsupportedPause(
+                int numRecordsPerSplitPerFetch,
+                boolean separatedFinishedRecord,
+                boolean blockingFetch) {
+            super(numRecordsPerSplitPerFetch, separatedFinishedRecord, blockingFetch);
+        }
+
+        @Override
+        public void pauseOrResumeSplits(
+                Collection<MockSourceSplit> splitsToPause,
+                Collection<MockSourceSplit> splitsToResume) {
+            throw new UnsupportedOperationException();
+        }
+
+        public static class Builder extends MockSplitReader.Builder {
+            public Builder(MockSplitReader.Builder other) {
+                super(other);
+            }
+
+            @Override
+            public MockSplitReader build() {
+                return new MockSplitReaderUnsupportedPause(
+                        numRecordsPerSplitPerFetch, separatedFinishedRecord, blockingFetch);
+            }
+        }
+
+        public static Builder cloneBuilder(MockSplitReader.Builder other) {
+            return new MockSplitReaderUnsupportedPause.Builder(other);
+        }
+    }
+
+    private static class SteppingSourceReaderTestHarness {
+        private final MockSteppingSplitFetcherManager<int[], MockSourceSplit> fetcherManager;
+        private final MockSourceReader sourceReader;
+
+        public SteppingSourceReaderTestHarness(
+                Supplier<SplitReader<int[], MockSourceSplit>> splitReaderSupplier,
+                Configuration configuration) {
+            FutureCompletingBlockingQueue<RecordsWithSplitIds<int[]>> queue =
+                    new FutureCompletingBlockingQueue<>(10);
+            this.fetcherManager =
+                    new MockSteppingSplitFetcherManager<>(
+                            queue, splitReaderSupplier, configuration);
+            this.sourceReader =
+                    new MockSourceReader(
+                            queue, fetcherManager, configuration, new TestingReaderContext());
+        }
+
+        private static List<MockSourceSplit> createPrefilledSplits(int numSplits, int numRecords) {
+            final List<MockSourceSplit> splits = new ArrayList<>(numSplits);
+            for (int splitId = 0; splitId < numSplits; splitId++) {
+                MockSourceSplit split = new MockSourceSplit(splitId, 0, numRecords);
+                for (int i = 0; i < numRecords; i++) {
+                    split.addRecord(i * numSplits + splitId);
+                }
+                splits.add(split);
+            }
+            return splits;
+        }
+
+        public void addPrefilledSplitsSingleReader(int numSplits, int numRecords) {
+            sourceReader.addSplits(createPrefilledSplits(numSplits, numRecords));
+            sourceReader.notifyNoMoreSplits();
```

Comment:
Is this a required assumption? We should also exercise the case where splits are dynamically discovered (e.g. Kafka source topic partition discovery).

Reply:
This is not required. The signal simply allows for a clean shutdown of the source.